packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1915 @@
+} from './Caveat';
+import {
+  constructPermission,
+  findCaveat,
+  GenericPermission,
+  OriginString,
+  PermissionOptions,
+  RequestedPermissions,
+  RestrictedMethod,
+  RestrictedMethodParameters,
+  ValidatePermissionSpecification,
+  PermissionSpecificationMap,
+  Permission,
+  ExtractAllowedCaveatTypes,
+  MutableGenericPermission,
+  ExtractPermissionSpecification,
+  PermissionSpecificationConstraint,
+} from './Permission';
+import {
+  PermissionDoesNotExistError,
+  methodNotFound,
+  UnrecognizedSubjectError,
+  CaveatDoesNotExistError,
+  UnrecognizedCaveatTypeError,
+  CaveatMissingValueError,
+  InvalidCaveatFieldsError,
+  InvalidSubjectIdentifierError,
+  InvalidCaveatTypeError,
+  CaveatAlreadyExistsError,
+  InvalidCaveatError,
+  InvalidApprovedPermissionError,
+  unauthorized,
+  invalidParams,
+  internalError,
+  userRejectedRequest,
+  PermissionsRequestNotFoundError,
+  ForbiddenCaveatError,
+} from './errors';
+import { MethodNames } from './utils';
+import { getPermissionMiddlewareFactory } from './permission-middleware';
+
+/**
+ * Metadata associated with {@link PermissionController} subjects.
+ */
+export type PermissionSubjectMetadata = {
+  origin: OriginString;
+};
+
+/**
+ * Metadata associated with permission requests.
+ */
+export type PermissionsRequestMetadata = PermissionSubjectMetadata & {
+  id: string;
+};
+
+/**
+ * Used for prompting the user about a proposed new permission.
+ * Includes information about the grantee subject, requested permissions, and
+ * any additional information added by the consumer.
+ *
+ * All properties except `permissions` are passed to any factories found for
+ * the requested permissions.
+ */
+export type PermissionsRequest = {
+  metadata: PermissionsRequestMetadata;
+  permissions: RequestedPermissions;
+  [key: string]: Json;
+};
+
+/**
+ * The name of the {@link PermissionController}.
+ */
+const controllerName = 'PermissionController';
+
+/**
+ * Permissions associated with a {@link PermissionController} subject.
+ */
+export type SubjectPermissions<Permissions extends GenericPermission> = Record<
+  Permissions['parentCapability'],
+  Permissions
+>;
+
+/**
+ * Permissions and metadata associated with a {@link PermissionController}
+ * subject.
+ */
+export type PermissionSubjectEntry<
+  SubjectPermission extends GenericPermission,
+> = {
+  permissions: SubjectPermissions<SubjectPermission>;
+} & PermissionSubjectMetadata;
+
+/**
+ * All subjects of a {@link PermissionController}.
+ */
+export type PermissionControllerSubjects<
+  SubjectPermission extends GenericPermission,
+> = Record<OriginString, PermissionSubjectEntry<SubjectPermission>>;
+
+// TODO:TS4.4 Enable compiler flags to forbid unchecked member access
+/**
+ * The state of a {@link PermissionController}.
+ */
+export type PermissionControllerState<Permissions> =
+  Permissions extends GenericPermission
+    ? {
+        subjects: PermissionControllerSubjects<Permissions>;
+      }
+    : never;
```

Comment:
Nit: This can be simplified

```suggestion
export type PermissionControllerState<Permissions extends GenericPermission> = {
  subjects: PermissionControllerSubjects<Permissions>;
};
```